test/util/dynamic/dynamic.go

```diff
@@ -0,0 +1,91 @@
+package dynamic
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// Client returns a client from a given unstructured object.
+// It can be used when creating an object from a yaml file.
+type Client interface {
+	GetClient(obj *unstructured.Unstructured) (ResourceClient, error)
+}
+
+type client struct {
+	dynamic.Interface
+	mapping meta.RESTMapper
+}
+
+func NewDynamicClient(kubeConfig *rest.Config) (Client, error) {
+	// Create a dynamic client
+	cli, err := dynamic.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a discovery mapper
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	groupResources, err := restmapper.GetAPIGroupResources(discoveryClient)
+	if err != nil {
+		return nil, err
+	}
+	mapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+
+	return &client{cli, mapper}, nil
+}
+
+func (d *client) getGVR(obj *unstructured.Unstructured) (*meta.RESTMapping, error) {
+	var gvk = obj.GroupVersionKind()
+	return d.mapping.RESTMapping(gvk.GroupKind(), gvk.Version)
+}
+
+// GetClient returns a non-namespaced or namespaced ResourceClient depending on a given object.
+func (d *client) GetClient(obj *unstructured.Unstructured) (ResourceClient, error) {
+	gvr, err := d.getGVR(obj)
+	if err != nil {
+		return nil, err
+	}
+	if obj.GetNamespace() != "" {
+		return &resourceClient{d.Resource(gvr.Resource).Namespace(obj.GetNamespace())}, nil
+	}
+	return &resourceClient{d.Resource(gvr.Resource)}, nil
+}
+
+// ResourceClient implements methods that can be used for *K8sObjectWithRetry helper functions.
+type ResourceClient interface {
+	Get(ctx context.Context, name string, options metav1.GetOptions) (*unstructured.Unstructured, error)
+	Create(ctx context.Context, obj *unstructured.Unstructured, options metav1.CreateOptions) (*unstructured.Unstructured, error)
+	Delete(ctx context.Context, name string, options metav1.DeleteOptions) error
+}
+
+type resourceClient struct {
+	client dynamic.ResourceInterface
+}
+
+func (d *resourceClient) Get(ctx context.Context, name string, options metav1.GetOptions) (*unstructured.Unstructured, error) {
+	// Get the object without the subresources argument to align with methods of other clients.
```

Comment:
I don't understand this comment. What other clients are you referring to?

Reply:
https://pkg.go.dev/k8s.io/client-go/dynamic#ResourceInterface
This is dynamic client interface.

For example,
https://pkg.go.dev/k8s.io/client-go@v0.30.3/kubernetes/typed/core/v1#ServiceInterface
This is service client interface.

Because of the difference, it doesn't fit to these types.
https://github.com/Azure/ARO-RP/blob/e5339b26b2fecb25993b533efc7f1fab40bb255f/test/e2e/helpers.go#L29-L32

Reply:
but the comments are a bit too redundant. I'll put the explanation somewhere else and remove those comments.